mlir/include/mlir/Dialect/GPU/IR/GPUOps.td

```diff
@@ -74,7 +74,20 @@ class GPU_IndexOp<string mnemonic, list<Trait> traits = []> :
     }]>,
     OpBuilder<(ins "::mlir::Type":$resultType, "::mlir::gpu::Dimension":$dimension), [{
       build($_builder, $_state, resultType, dimension, /*upperBound=*/nullptr);
+    }]>,
+    OpBuilder<(ins "::mlir::gpu::Dimension":$dimension, "std::optional<int64_t>":$upperBound), [{
```

Comment:
Why is it optional?

Reply:
So that you can explicitly pass in `std::nullopt` to leave the bound unset

Reply:
I also don't get what convenience does it bring. Why would you demand a nullopt when there can be two builders: with and without the attribute/whatever? Or if we want a single builder - a default value.

Reply:
`upper_bound` is a fundamentally optional property.

This is designed to make people's lives easier when they're in possession of something like an `ArrayRef<int64_t> blockSizes`.

I'll note we already have the attribute-less builder.

The advantage to having this be optional is that there's downstream code that either optionally knows the block sizes or optionally wants to forget about them.